### ISSUE TITLE
Delete Roles corrupts positions

### ISSUE THREAD

**Issue by firynth** [hydrabolt/discord.js#1155]
I retweaked my code so that deletion of unused roles locked out everything else from running that could possibly affect the roles on a server.

Delete roles, by itself, is causing the positions for the roles to get corrupted so that one position ends up with a value that's higher than what is supposed to be the highest position role on the server, causing the Discord API to issue a "Bad Request" error when trying to make changes afterwards.

This is the code:
```javascript
function deleteRoles (roles, counter, deleteCallback = function () {}) {
    deletingRoles = true;
    if (roles.length > 0) {
        roles[counter].delete()
        .then(function () {
            if ((counter + 1) < roles.length) {
                deleteRoles(roles, counter + 1, deleteCallback);
            } else {
                deleteCallback();
                deletingRoles = false;
            }
        })
        .catch(function (err) {
            console.log("Error deleting roles (" + counter + ") " + err);
            deletingRoles = false;
        });
    } else {
        deleteCallback();
        deletingRoles = false;
    }
}
```
I don't know if this is on Discord's side or via this API, but it can't be fixed unless you manually move one of the positions.

**Comment by Programmix** [comment]
[unrelated to issue] I'd recommend using Promises rather than callbacks for your code. 😄

**Comment by firynth** [comment]
Callback is what function I want to run next after the delete promise is returned. I could have made separate functions for each of them, but Javascript lets me do this, so it's easier to do.


Related to the problem at hand. I just figured out how to update discord.js to version 11 with npm (it was being a butt and I had to just install it again). I'll let you guys know if it happens again.

**Comment by GusCaplan** [comment]
```js
function deleteRoles (roles, counter, deleteCallback = function () {}) {
    deletingRoles = true;
    if (roles.length > 0) {
        roles[counter].delete()
        .then(function () {
            if ((counter + 1) < roles.length) {
                deleteRoles(roles, counter + 1, deleteCallback);
            } else {
                deleteCallback();
                deletingRoles = false;
            }
        })
        .catch(function (err) {
            console.log("Error deleting roles (" + counter + ") " + err);
            deletingRoles = false;
        });
    } else {
        deleteCallback();
        deletingRoles = false;
    }
}

**Comment by Programmix** [comment]
```
function deleteRoles (roles) {
    const promises = [];
    
    for (const role of roles) {
      promises.push(role.delete());
    }

    return promises;
}```

This is a more sane way of doing what you're trying to accomplish. Discord.js's sequential sender will take care of the API request sending rate.

**Comment by firynth** [comment]
So you're making me handle an array of promises instead now, which I would have to handle with .all() on every element in the array in the previous function, which would also mean making another function to handle all promise returns of the array than doing it iteratively with one function, All instances of this function then can run async with one another and the whole reason why I coded it this way to be iterative in order to try to prevent the bug that is the subject of this thread? Because as soon as it occurs, it requires manual intervention to fix the problem.

If it's fixed with a recent update, then fine because I didn't know discord.js had a sequential sender. But that function was coded originally to occur after stripping a user of one or more roles, which passes it's callback function to this function, then delete any roles with no members in them BEFORE creating a new role or trying to add an existing one to a user.

I get the array being returned should accomplish this, but I'm supposed to wait for every single instance to be finished before doing *anything* else and if anything errors out, I want to know which one caused it and stop the code from screwing up the server roles anymore than originally happens. FYI, your function would be nicer since it never seems to actually error there, only when I try to change positions, but you're recommending I do something without knowing why I spent hours reworking my original code to this setup in the first place.
